### PR TITLE
Adds offline muting

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Muting is now be processed client-side, for offline users. To run transitions on the client, no config changes are required. However, it is possible to disable client-side muting by updating the config like: 
```json
"transitions": {
  "muting": { "client_side": false }
}
```

Also: 
Splits validation into own shared lib to be used from webapp.
Updates validation to be promise based (some QOL but kept changes to a minimum). 
Updates lineage to use all known docs when building subject lineages (so when we can have hydration for new reports + new contacts out of the box). 
Consolidates the get-message logic into message-utils to be used by validation and transitions.
Consolidates getLocale into message-utils.

Offline muting happens on the device, with the transition code running over docs before they are saved into the local database, following a few rules:
1. Only new docs are processed
2. When a batch contains both muting and unmuting forms, only unmuting forms will be processed, while the muting forms will be ignored.
3. The actions performed over docs are: 
 - when a muting / unmuting report is submitted, local relevant contacts are updated. This follows the same rules as server-side muting, where muting cascades downwards and unmuting cascades upwards.
 - when a new contact is added under a muted parent, it also becomes muted
 - reports get an "client_side_transitions" property to log which transitions have run over them
 - validation runs over reports, with invalid reports being ignored by the transition and validation errors saved in their json. Validation errors are generated in the user's selected language and are visible on the report's detail page. 
 - contacts get a "muting_history" property, to log every muting/unmuting event that has been processed on the device, and which "service" updated the muted state last (options are "server_side" or "client_side")
 - the "muted" property on the contact is still the property that determines if the contact is muted or not, so configs and queries don't require changes. 
4. Except the muting/unmuting reports and the relevant contacts (as described above), no other docs are edited directly -  registrations are not silenced and the config is still responsible with handling muted contacts (cancelling tasks, updating targets. contact summary and etc). 
5. It's necessary to process docs before they are saved, otherwise there is a high risk of generating conflicts on every muting event, if the device has a strong internet connection and Sentinel is up to date (and both device and Sentinel process the event at the same time). 

Server-side muting is updated to handle offline muting:
1. When a contact is processed, its "muting_history.last_update" is updated to "server_side" and an entry is added to the infodoc muting_history.
   If the contact is muted, the muted property is updated with the server's current date (for consistency).
   The contact's offline "muting_history" event log is left untouched and is **not** merged into the infodoc muting_history (to avoid adding complex logic around handling doc conflicts and ending up with muting_history events stored in infodocs that happened only for losing revs).
2. Muting/unmuting reports that were processed offline are always processed sever-side, even if their subjects are already in the correct state (because offline muting doesn't silence registrations, and offline devices don't have access to the whole hierarchy to perform correct muting).
   Additionally, for every contact that is updated, the offline "muting_history" event log is inspected and searched for the muting event corresponding to the currently processed report. If a matching event is found, every other following event is collected to be replayed. 
   After processing all contacts, the collected list of muting events to be replayed is sorted chronologically (this sorting is imperfect, as dates from devices are not reliable, but some sorting is necessary and this is the only data that is available). This results in a sorted list of report uuids. Muting is (re)run over every corresponding report, in sequence and in order, even if muting already ran. 
   Replaying events that were processed on the client (ideally in the same order they were submitted) is necessary because there are little chances for Sentinel to process these events in this ideal order (mainly because the changes feed is not ordered and the position of a doc in the feed is not correlated with when it was created, rather with when it was updated/synced/read from a shard).
3. When receiving a contact that was last_updated offline, Sentinel processes it - to solve the problem of the muting/unmuting report being deleted before it was synced. 
   

The code tries to handle most situations that might arise, in an effort to make the muting action idempotent:
- a composite form that creates a new place, + new persons in the place and mutes this new place (which should mute the new place and the new persons in the place)
- a composite form that creates a new place + new persons in the place and mutes the person (which should only mute the new person)
- a composite form that creates a new person under a muted parent and unmutes the person or any other ancestor (which should unmute all contacts under the topmost muted ancestor). 
- multiple muting/unmuting events processed on the client and then synced out of order. 
- deletion of muting/unmuting reports before syncing

However, there could be situations where muting client-side could yield inconsistent results: for example if settings are updated, so a device and Sentinel run muting over the same doc with different settings, could result in contacts being muted but registrations not silenced. Additionally, the behavior might not be consistent when reports are edited before they are synced (and one contact is muted on the client-side and another contact is muted server-side). 

medic/cht-core#6737

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: shared-libs/transitions/src/lib/muting_utils.js [hunk 1]
```diff
@@ -8,33 +8,7 @@ infodoc.initLib(db.medic, db.sentinel);
...
-      return result;
-    })
-    .then(result => {
-      if (!result.rows.length) {
-        throw(new Error('contact_not_found'));
-      }
-      return lineage.fetchHydratedDoc(result.rows[0].id);
-    });
-};
+const getContact = doc => doc.patient || doc.place;
```
Comment (dianabarsan) [inline shared-libs/transitions/src/lib/muting_utils.js:11]: Place subjects are now hydrated out of the box!

File: shared-libs/validation/src/validation.js [hunk 2]
```diff
@@ -0,0 +1,316 @@
+const _ = require('lodash/core');
```
Comment (dianabarsan) [inline shared-libs/validation/src/validation.js:1]: I moved this file but GH still reports full edits :( 
I tried to keep changes to a minimum, but now that it's 100% changed, I'm wondering if we should prettify it even more.

**Comment by dianabarsan** [comment]
hey @garethbowen . If you don't mind, I'd ask you to have a look, please. 

This is a precursor to the muting work, by making some code usable by webapp (like validation) and updating lineage to support more robust hydration when passing arrays of "new" contacts. 
My initial idea was to try to make the whole `transitions` shared lib to be available in webapp, but I abandoned it rather quickly. 

My thoughts were that, once this is approved, I'll rebase the main muting work branch and have that approved, merge the work into this branch and have it AT-ed together. 

Thanks for your time!

File: shared-libs/message-utils/src/index.js [hunk 3]
```diff
@@ -345,26 +345,51 @@ exports.generate = function(config, translate, doc, content, recipient, extraCon
...
+  }
+  // use the translation key if provided
+  if (configuration.translation_key) {
+    return translate(configuration.translation_key, locale);
+  }
+
+  // use the translation key if provided
+  if (configuration.translationKey) {
+    return translate(configuration.translationKey, locale);
+  }
```
Comment (garethbowen) [inline shared-libs/message-utils/src/index.js:376]: Combine these branches, eg:

```suggestion
  // use the translation key if provided
  const translationKey = configuration.translation_key || configuration.translationKey;
  if (translationKey) {
    return translate(translationKey, locale);
  }
```

File: shared-libs/transitions/src/lib/muting_utils.js (hunk 1 above)
Comment (garethbowen) [reply shared-libs/transitions/src/lib/muting_utils.js:11]: This seems too trivial to export now - is it worth inlining this wherever it's used to make it obvious what it does?

File: shared-libs/transitions/src/transitions/utils.js [hunk 4]
```diff
@@ -88,6 +92,6 @@ module.exports = {
 
   validate: (config, doc) => {
     const validations = config && config.validations && config.validations.list;
-    return new Promise(resolve => validation.validate(doc, validations, resolve));
+    return validation.validate(doc, validations);
```
Comment (garethbowen) [inline shared-libs/transitions/src/transitions/utils.js:95]: So nice!

**Review by garethbowen — APPROVED** [review]
So much nicer! 2 small improvements suggested but approved so you're not blocked.

File: shared-libs/transitions/src/lib/muting_utils.js (hunk 1 above)
Comment (dianabarsan) [reply shared-libs/transitions/src/lib/muting_utils.js]: Yeap, but I had a bunch of unit tests to change :D

**Pull request merged**